File: server/apps/insights/groq.py
"""Groq inference for weekly AI insights. Plain urllib, matching the rest of
this codebase's external-API style (apps/devices/geoip.py,
apps/accounts/tg_api.py) — Groq's API is a bare OpenAI-shaped JSON endpoint,
so no client library is worth adding for one call site.

⚠️ GROQ_MODEL is not hardcoded here — Groq's lineup and pricing change, and
this project doesn't maintain a pricing/model reference for them. Confirm
the current model id and cost at https://console.groq.com before setting
GROQ_MODEL in production.

⚠️ PythonAnywhere's outbound IP range is blocked by Cloudflare in front of
api.groq.com (HTTP 403, "error code: 1010" — an ASN/datacenter block, not
anything wrong with the request). Confirmed by hand from a PA console.
Calls are routed through a tiny relay on Vercel instead
(parent-web/src/app/api/groq-relay/route.ts) when GROQ_RELAY_URL is set —
Vercel's IPs aren't on that block list. GROQ_RELAY_SECRET authenticates to
the relay; the real GROQ_API_KEY only needs to exist wherever the actual
call happens (the relay when routed through it, this process otherwise).
"""

import json
import logging
import urllib.request

from django.conf import settings

logger = logging.getLogger(__name__)

_TIMEOUT = 30
_DIRECT_API_URL = "https://api.groq.com/openai/v1/chat/completions"

_SYSTEM_PROMPT = (
    "Siz Spino24 — bolalar uchun raqamli xavfsizlik dasturidagi yordamchisiz. "
    "Ota-onaga farzandining haftalik ekran faoliyati haqida qisqa, tushunarli va "
    "hukm qilmaydigan tahlil bering. Har doim o'zbek tilida, iliq va hamkorlik "
    "ruhida yozing — bolani ayblamang, ota-onaga amaliy tavsiyalar bering.\n\n"
    "Faqat quyidagi JSON formatida javob bering, boshqa hech narsa yozmang:\n"
    '{"summary": "2-3 gapli umumiy holat", '
    '"highlights": ["diqqatga molik narsa 1", "..."], '
    '"recommendations": ["amaliy tavsiya 1", "..."], '
    '"risk_level": "ok" | "watch" | "concern"}'
)


def is_configured() -> bool:
    # Either a direct key (this process calls Groq itself) or a relay URL +
    # secret (the relay holds the real key) is enough — GROQ_MODEL is always
    # required since it's sent in the request body either way.
    has_credentials = bool(settings.GROQ_API_KEY) or bool(
        settings.GROQ_RELAY_URL and settings.GROQ_RELAY_SECRET
    )
    return bool(has_credentials and settings.GROQ_MODEL)


def _build_user_prompt(week_data: dict) -> str:
    return (
        f"Farzand: {week_data['child_name']}\n"
        f"Hafta: {week_data['week_start']} — {week_data['week_end']}\n"
        f"Jami ekran vaqti: {week_data['total_minutes']} daqiqa\n"
        f"Kunlar bo'yicha (daqiqa): {json.dumps(week_data['daily_minutes'], ensure_ascii=False)}\n"
        f"Eng ko'p ishlatilgan ilovalar: {json.dumps(week_data['top_apps'], ensure_ascii=False)}\n"
        f"Eng ko'p tashrif buyurilgan saytlar: {json.dumps(week_data['top_domains'], ensure_ascii=False)}\n"
        f"Ogohlantirishlar (turi -> soni): {json.dumps(week_data['alerts'], ensure_ascii=False)}\n"
        f"O'rnatilgan qoidalar: {json.dumps(week_data['rules'], ensure_ascii=False)}\n"
    )


def generate_weekly_insight(week_data: dict) -> dict | None:
    """Calls Groq and returns a dict matching WeeklyInsight's fields
    (summary/highlights/recommendations/risk_level), or None on any
    failure — never raises, mirrors every other best-effort external call
    in this codebase."""
    if not is_configured():
        return None

    body = {
        "model": settings.GROQ_MODEL,
        "messages": [
            {"role": "system", "content": _SYSTEM_PROMPT},
            {"role": "user", "content": _build_user_prompt(week_data)},
        ],
        "temperature": 0.4,
        "response_format": {"type": "json_object"},
    }

    if settings.GROQ_RELAY_URL and settings.GROQ_RELAY_SECRET:
        url = settings.GROQ_RELAY_URL
        headers = {"X-Relay-Secret": settings.GROQ_RELAY_SECRET, "Content-Type": "application/json"}
    else:
        url = _DIRECT_API_URL
        headers = {"Authorization": f"Bearer {settings.GROQ_API_KEY}", "Content-Type": "application/json"}

    request = urllib.request.Request(url, data=json.dumps(body).encode(), headers=headers)
    try:
        with urllib.request.urlopen(request, timeout=_TIMEOUT) as response:
            data = json.loads(response.read())
        content = data["choices"][0]["message"]["content"]
        parsed = json.loads(content)
    except Exception:
        logger.exception("Groq weekly insight generation failed")
        return None

    summary = str(parsed.get("summary") or "").strip()
    if not summary:
        return None
    risk_level = parsed.get("risk_level")
    if risk_level not in ("ok", "watch", "concern"):
        risk_level = "ok"
    return {
        "summary": summary,
        "highlights": [str(h) for h in (parsed.get("highlights") or [])][:10],
        "recommendations": [str(r) for r in (parsed.get("recommendations") or [])][:10],
        "risk_level": risk_level,
    }

File: parent-web/src/lib/groq.ts
// Groq inference for "AI tahlil" weekly insights — lives here (Vercel), not
// in the Django backend, because PythonAnywhere Free's outbound internet is
// restricted to a fixed whitelist that api.groq.com's Cloudflare front-end
// rejects by IP/ASN even when PA's own proxy lets the connection through
// (confirmed by hand: HTTP 403 "error code: 1010"). Vercel's egress isn't
// affected, so every Groq call — on-demand (via /api/groq-relay, called by
// the client) and proactive weekly (via /api/weekly-insight-cron, called by
// an external cron) — happens from here instead.

export type RiskLevel = "ok" | "watch" | "concern";
export type Period = "last_week" | "this_week" | "today";

export type WeekData = {
  child_name: string;
  week_start: string;
  week_end: string;
  total_minutes: number;
  daily_minutes: Record<string, number>;
  top_apps: { name: string; minutes: number }[];
  top_domains: { domain: string; visits: number }[];
  alerts: Record<string, number>;
  rules: { rule_type: string; value: unknown }[];
  period: Period;
  // this_week/today are necessarily still in progress — the prompt must
  // say so, or the model narrates an unfinished period as a finished one
  // ("bu hafta jami X daqiqa sarfladi" about a week that's only half over).
  is_partial: boolean;
};

export type InsightResult = {
  summary: string;
  highlights: string[];
  recommendations: string[];
  risk_level: RiskLevel;
};

const SYSTEM_PROMPT = `Siz Spino24 — bolalar uchun raqamli xavfsizlik dasturidagi yordamchisiz. Ota-onaga farzandining ekran faoliyati haqida qisqa, tushunarli va hukm qilmaydigan tahlil bering. Har doim o'zbek tilida, iliq va hamkorlik ruhida yozing — bolani ayblamang, ota-onaga amaliy tavsiyalar bering.

Tahlil qilinayotgan davr hali tugamagan bo'lishi mumkin (masalan, "shu hafta" yoki "bugun") — bunday holda buni tugallangan davrdek emas, hali davom etayotgan davr sifatida tasvirlang (masalan, "shu haftaning hozirgacha bo'lgan qismida" kabi), va son/foizlarni "hozirgacha" degan ma'noda bering.

Faqat quyidagi JSON formatida javob bering, boshqa hech narsa yozmang:
{"summary": "2-3 gapli umumiy holat", "highlights": ["diqqatga molik narsa 1", "..."], "recommendations": ["amaliy tavsiya 1", "..."], "risk_level": "ok" | "watch" | "concern"}`;

const PERIOD_LABEL: Record<Period, string> = {
  last_week: "O'tgan hafta (to'liq tugagan)",
  this_week: "Shu hafta (hali tugamagan, dushanbadan bugungi kungacha)",
  today: "Bugun (kun hali davom etmoqda)",
};

function buildUserPrompt(week: WeekData): string {
  return [
    `Farzand: ${week.child_name}`,
    `Davr: ${PERIOD_LABEL[week.period] ?? week.period} — ${week.week_start} — ${week.week_end}`,
    week.is_partial ? "Diqqat: bu davr hali tugamagan, ma'lumot faqat hozirgacha bo'lgan qismni qamrab oladi." : "",
    `Jami ekran vaqti: ${week.total_minutes} daqiqa`,
    `Kunlar bo'yicha (daqiqa): ${JSON.stringify(week.daily_minutes)}`,
    `Eng ko'p ishlatilgan ilovalar: ${JSON.stringify(week.top_apps)}`,
    `Eng ko'p tashrif buyurilgan saytlar: ${JSON.stringify(week.top_domains)}`,
    `Ogohlantirishlar (turi -> soni): ${JSON.stringify(week.alerts)}`,
    `O'rnatilgan qoidalar: ${JSON.stringify(week.rules)}`,
  ]
    .filter(Boolean)
    .join("\n");
}

/** Calls Groq directly (this function only ever runs on Vercel — the
 * weekly cron route). Returns null on any failure; never throws, matching
 * the rest of this project's "external calls are best-effort" posture. */
export async function generateWeeklyInsight(week: WeekData): Promise<InsightResult | null> {
  const apiKey = process.env.GROQ_API_KEY;
  const model = process.env.GROQ_MODEL;
  if (!apiKey || !model) return null;

  try {
    const res = await fetch("https://api.groq.com/openai/v1/chat/completions", {
      method: "POST",
      headers: { Authorization: `Bearer ${apiKey}`, "Content-Type": "application/json" },
      body: JSON.stringify({
        model,
        messages: [
          { role: "system", content: SYSTEM_PROMPT },
          { role: "user", content: buildUserPrompt(week) },
        ],
        temperature: 0.4,
        response_format: { type: "json_object" },
      }),
    });
    if (!res.ok) return null;
    const data = await res.json();
    const parsed = JSON.parse(data.choices[0].message.content);
    return normalizeResult(parsed);
  } catch {
    return null;
  }
}

export function normalizeResult(parsed: any): InsightResult | null {
  const summary = String(parsed?.summary ?? "").trim();
  if (!summary) return null;
  const riskLevel: RiskLevel = ["ok", "watch", "concern"].includes(parsed?.risk_level)
    ? parsed.risk_level
    : "ok";
  return {
    summary,
    highlights: Array.isArray(parsed?.highlights) ? parsed.highlights.map(String).slice(0, 10) : [],
    recommendations: Array.isArray(parsed?.recommendations)
      ? parsed.recommendations.map(String).slice(0, 10)
      : [],
    risk_level: riskLevel,
  };
}

export { SYSTEM_PROMPT, buildUserPrompt };
